<div style='background-color: #e3a8b6;
    border: 0.5em solid black;
    border-radius: 0.5em;
    padding: 1em;'>
    <h2>Exercice</h2>
    <h1>Composition des produits alimentaires</h1>
</div>

L'API [_Open Food Facts_](https://fr.openfoodfacts.org/) permet de récupérer des informations sur un produit alimentaire à partir de son numéro de code-barres. Par exemple, pour le produit dont le code-barres est `8076800376999`, il suffit d'envoyer une requête GET à l'URL [https://world.openfoodfacts.org/api/v0/product/8076800376999.json](https://world.openfoodfacts.org/api/v0/product/8076800376999.json).

**(1)** ✏️ 💻 Après avoir testé la fonction `interroger_API_openfoodfacts` pour le code-barres `8076800376999`, écrire sa spécification.

In [1]:
import requests

In [2]:
def interroger_API_openfoodfacts(code_barres):
    """
    Retourne, à partir d'un code-barres, le nom, le nutri score et les nutriments contenus dans un produit.
    - Entrée : code_barres (entier ou chaîne de caractères)
    - Sortie : dictionnaire contenant les clés 'nom', 'nutriments' et 'nutri_score' (les deux dernières
               clés uniquement si le code_barres est reconnu par l'API)
    """
    reponse = requests.get(f"https://world.openfoodfacts.org/api/v0/product/{code_barres}.json")
    reponse = reponse.json()
    if reponse['status'] == 1:
        dico = {'nom': reponse['product']['product_name'],
                'nutriments': reponse['product']['nutriments'],
                'nutri_score': reponse['product']['nutrition_grade_fr']}
    else:
        dico = {'nom': 'Produit inconnu'}
    return dico

In [3]:
interroger_API_openfoodfacts(8076800376999)

{'nom': "Lasagne all'uovo",
 'nutriments': {'carbohydrates': 67.3,
  'carbohydrates_100g': 67.3,
  'carbohydrates_unit': 'g',
  'carbohydrates_value': 67.3,
  'carbon-footprint-from-known-ingredients_100g': 50.336,
  'carbon-footprint-from-known-ingredients_product': 252,
  'energy': 1554,
  'energy-kcal': 367,
  'energy-kcal_100g': 367,
  'energy-kcal_unit': 'kcal',
  'energy-kcal_value': 367,
  'energy-kj': 1554,
  'energy-kj_100g': 1554,
  'energy-kj_unit': 'kJ',
  'energy-kj_value': 1554,
  'energy_100g': 1554,
  'energy_unit': 'kJ',
  'energy_value': 1554,
  'fat': 4,
  'fat_100g': 4,
  'fat_unit': 'g',
  'fat_value': 4,
  'fiber': 3,
  'fiber_100g': 3,
  'fiber_unit': 'g',
  'fiber_value': 3,
  'fruits-vegetables-nuts': 0,
  'fruits-vegetables-nuts-estimate-from-ingredients_100g': 0,
  'fruits-vegetables-nuts_100g': 0,
  'fruits-vegetables-nuts_serving': 0,
  'fruits-vegetables-nuts_unit': 'g',
  'fruits-vegetables-nuts_value': 0,
  'nova-group': 1,
  'nova-group_100g': 1,
  'nov

Voici un extrait du tableau d'information nutritionnelle d'un paquet de lasagnes dont le code-barres est `8076800376999` :

<table>
    <tr>
        <th style='text-align: center;'>Information nutritionnelle</th>
        <th style='text-align: center;'>Pour 100 g</th>
    </tr>
    <tr>
        <td style='text-align: left;'>Énergie</td>
        <td style='text-align: center;'>1554 kJ</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Matières grasses</td>
        <td style='text-align: center;'>4 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'><i>dont acides gras saturés</i></td>
        <td style='text-align: center;'>1,2 g</td>
    </tr>
        <tr>
        <td style='text-align: left;'>Glucides</td>
        <td style='text-align: center;'>67,3 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'><i>dont sucres</i></td>
        <td style='text-align: center;'>3 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Fibres alimentaires</td>
        <td style='text-align: center;'>3 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Protéines</td>
        <td style='text-align: center;'>14 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Sel</td>
        <td style='text-align: center;'>0,075 g</td>
    </tr>
</table>

**(2)** ✏️ Déterminer comment accéder à la quantité d'énergie, de sel, de sucres, de matières grasses, de fibres et de protéines à partir de la réponse fournie par l'API.

In [4]:
donnees = interroger_API_openfoodfacts(8076800376999)
print(f"Quantité d'énergie ......... : {donnees['nutriments']['energy']} kJ")
print(f"         de sel ............ : {donnees['nutriments']['salt']} g")
print(f"         de sucres ......... : {donnees['nutriments']['sugars']} g")
print(f"         de matières grasses : {donnees['nutriments']['fat']} g")
print(f"         de fibres ......... : {donnees['nutriments']['fiber']} g")
print(f"         de protéines ...... : {donnees['nutriments']['proteins']} g")

Quantité d'énergie ......... : 1554 kJ
         de sel ............ : 0.075 g
         de sucres ......... : 3 g
         de matières grasses : 4 g
         de fibres ......... : 3 g
         de protéines ...... : 14 g


**(3)** ✏️ 💻 Écrire la spécification de la fonction `mystere`.

In [5]:
def mystere(dico, cle):
    """
    Retourne dico[cle] si cette valeur existe, et 0 sinon.
    - Entrées : dico (dictionnaire), cle (clé potentielle du dictionnaire dico)
    - Sortie : valeur associée à cle dans le dictionnaire si elle existe, 0 sinon
    """
    if cle in dico:
        return dico[cle]
    else:
        return 0

**(4)** 💻 Définir une fonction `composition` prenant en paramètre d'entrée un numéro de code-barres et retournant un dictionnaire dont les clés sont :
- `'nom'` : nom du produit,
- `'énergie'` : énergie pour 100g de produit, exprimée en kJ,
- `'sel'` : quantité de sel pour 100g de produit, exprimée en g,
- `'sucres'` : quantité de sucres pour 100g de produit, exprimée en g,
- `'mat_grasses'` : quantité de matières grasses pour 100g de produit, exprimée en g,
- `'fibres'` : quantité de fibres pour 100g de produit, exprimée en g,
- `'protéines'` : quantité de protéines pour 100g de produit, exprimée en g.

In [6]:
def composition(code_barres):
    """
    Détermine la quantité de certains nutriments dans un produit alimentaire, à partir de son code-barres.
    - Entrée : code_barres (entier ou chaîne de caractères)
    - Sortie : dictionnaire contenant les clés 'nom', 'énergie', 'sel', 'sucres', 'mat_grasses', 'fibres', 'protéines'
    """
    donnees = interroger_API_openfoodfacts(code_barres)
    nom = donnees['nom']
    energie = mystere(donnees['nutriments'], 'energy')
    sel = mystere(donnees['nutriments'], 'salt')
    sucres = mystere(donnees['nutriments'], 'sugars')
    mat_grasses = mystere(donnees['nutriments'], 'fat')
    fibres = mystere(donnees['nutriments'], 'fiber')
    proteines = mystere(donnees['nutriments'], 'proteins')
    return {'nom' : nom, 'énergie' : energie, 'sel' : sel, 'sucres' : sucres,
            'mat_grasses' : mat_grasses, 'fibres' : fibres, 'protéines' : proteines}

In [7]:
composition(8076800376999)

{'nom': "Lasagne all'uovo",
 'énergie': 1554,
 'sel': 0.075,
 'sucres': 3,
 'mat_grasses': 4,
 'fibres': 3,
 'protéines': 14}

**(5)** 💻 Définir une procédure `comparer_produits` qui prend en paramètre d'entrée un tableau contenant des numéros de code-barres et qui affiche :
- le produit qui contient le plus d'énergie pour 100g,
- le produit qui contient le moins de sel pour 100g,
- le produit qui contient le moins de matières grasses pour 100g,
- le produit qui contient le plus de sucres pour 100g.

In [8]:
def _recherche_min(prod, cle):
    """
    Détermine l'indice de l'élément de prod qui a la plus petite valeur associée à une clé.
    - Entrées : prod (tableau de dictionnaires), cle (cle des dictionnaires éléments de prod)
    - Sortie : indice_min (entier)
    """
    indice_min = 0
    for k in range(1, len(prod)):
        if prod[k][cle] < prod[indice_min][cle]:
            indice_min = k
    return indice_min

def _recherche_max(prod, cle):
    """
    Détermine l'indice de l'élément de prod qui a la plus grande valeur associée à une clé.
    - Entrées : prod (tableau de dictionnaires), cle (cle des dictionnaires éléments de prod)
    - Sortie : indice_max (entier)
    """
    indice_max = 0
    for k in range(1, len(prod)):
        if prod[k][cle] > prod[indice_max][cle]:
            indice_max = k
    return indice_max
    
def comparer_produits(tab):
    """
    Affiche le produit contenant le plus d'énergie, le moins de sel, le moins de gras, le plus de sucres.
    - Entrée : tab (tableau dont les éléments sont des numéros de code-barres)
    - Effet de bord : affichage à l'écran
    """
    produits = [composition(code_barres) for code_barres in tab]
    print(f"Produit contenant le plus d'énergie : {produits[_recherche_max(produits, 'énergie')]['nom']}")
    print(f"Produit contenant le moins de sel : {produits[_recherche_min(produits, 'sel')]['nom']}")
    print(f"Produit contenant le moins de matières grasses : {produits[_recherche_min(produits, 'mat_grasses')]['nom']}")
    print(f"Produit contenant le plus de sucres : {produits[_recherche_max(produits, 'sucres')]['nom']}")

Vous pouvez tester votre fonction avec les produits suivants :

<table>
    <tr>
        <th style='text-align: left;'>Nom du produit</th>
        <th style='text-align: center;'>Granola</th>
        <th style='text-align: center;'>Petits pois Carottes</th>
        <th style='text-align: center;'>Paëlla</th>
        <th style='text-align: center;'>Ice tea pêche</th>
        <th style='text-align: center;'>Blanc de poulet</th>
        <th style='text-align: center;'>Cassoulet</th>
        <th style='text-align: center;'>Coquillettes</th>
    </tr>
    <tr>
        <td style='text-align: left;'>Code-barres</td>
        <td style='text-align: center;'>7622300689124</td>
        <td style='text-align: center;'>3083680026321</td>
        <td style='text-align: center;'>3302741859105</td>
        <td style='text-align: center;'>5449000232465</td>
        <td style='text-align: center;'>3095756193011</td>
        <td style='text-align: center;'>3261055930422</td>
        <td style='text-align: center;'>8076808140325</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Energie</td>
        <td style='text-align: center;'>2108 kJ</td>
        <td style='text-align: center;'>226 kJ</td>    
        <td style='text-align: center;'>888 kJ</td>
        <td style='text-align: center;'>79 kJ</td>
        <td style='text-align: center;'>416 kJ</td>
        <td style='text-align: center;'>480 kJ</td>
        <td style='text-align: center;'>1521 kJ</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Matières grasses</td>
        <td style='text-align: center;'>26 g</td>
        <td style='text-align: center;'>0,5 g</td>
        <td style='text-align: center;'>10 g</td>
        <td style='text-align: center;'>0 g</td>
        <td style='text-align: center;'>1,6 g</td>
        <td style='text-align: center;'>4,7 g</td>
        <td style='text-align: center;'>2 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'><i>dont acides gras saturés</i></td>
        <td style='text-align: center;'>14 g</td>
        <td style='text-align: center;'>0,1 g</td>
        <td style='text-align: center;'>2,8 g</td>
        <td style='text-align: center;'>0 g</td>
        <td style='text-align: center;'>0,4 g</td>
        <td style='text-align: center;'>1,8 g</td>
        <td style='text-align: center;'>0,5 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Glucides</td>
        <td style='text-align: center;'>61 g</td>
        <td style='text-align: center;'>7,2 g</td>
        <td style='text-align: center;'>16 g</td>
        <td style='text-align: center;'>4,3 g</td>
        <td style='text-align: center;'>0,5 g</td>
        <td style='text-align: center;'>8,9 g</td>
        <td style='text-align: center;'>71,2 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'><i>dont sucres</i></td>
        <td style='text-align: center;'>36 g</td>
        <td style='text-align: center;'>3,4 g</td>
        <td style='text-align: center;'>0,6 g</td>
        <td style='text-align: center;'>4,3 g</td>
        <td style='text-align: center;'>0,5 g</td>
        <td style='text-align: center;'>0,6 g</td>
        <td style='text-align: center;'>3,5 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Fibres alimentaires</td>
        <td style='text-align: center;'>2,9 g</td>
        <td style='text-align: center;'>4,5 g</td>
        <td style='text-align: center;'>1,2 g</td>
        <td style='text-align: center;'>0 g</td>
        <td style='text-align: center;'>0 g</td>
        <td style='text-align: center;'>3,8 g</td>
        <td style='text-align: center;'>3 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Protéines</td>
        <td style='text-align: center;'>4,5 g</td> 
        <td style='text-align: center;'>2,9 g</td>
        <td style='text-align: center;'>10 g</td>
        <td style='text-align: center;'>0 g</td>
        <td style='text-align: center;'>21 g</td>
        <td style='text-align: center;'>7,3 g</td>
        <td style='text-align: center;'>12,5 g</td>
    </tr>
    <tr>
        <td style='text-align: left;'>Sel</td>
        <td style='text-align: center;'>1,05 g</td>
        <td style='text-align: center;'>0,57 g</td>
        <td style='text-align: center;'>0,8 g</td>
        <td style='text-align: center;'>0,03 g</td>
        <td style='text-align: center;'>1,8 g</td>
        <td style='text-align: center;'>0,81 g</td>
        <td style='text-align: center;'>0,013 g</td>
    </tr>
</table>

In [9]:
liste = [7622300689124, 3083680026321, 3302741859105, 5449000232465, 3095756193011, 3261055930422, 8076808140325]
comparer_produits(liste)

Produit contenant le plus d'énergie : Granola Gros éclats de Chocolat
Produit contenant le moins de sel : Barilla pates coquillettes
Produit contenant le moins de matières grasses : FuzeTea Thé glacé pêche intense
Produit contenant le plus de sucres : Granola Gros éclats de Chocolat
